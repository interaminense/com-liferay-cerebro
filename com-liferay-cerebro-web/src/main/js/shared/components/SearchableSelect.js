import autobind from 'autobind-decorator';
import Component, {Config} from 'metal-jsx';
import {noop} from 'lodash';

import Button from './Button';
import Dropdown from './Dropdown';
import SearchInput from './SearchInput';

class Item extends Component {
	@autobind
	handleClick() {
		this.props.onSelect(this.props.item);
	}

	render() {
		const {item} = this.props;

		return (
			<Dropdown.Item
				{...this.otherProps()}
				hideOnClick
				onClick={this.handleClick}
			>
				{item.name}
			</Dropdown.Item>
		);
	}
}

Item.PROPS = {
	item: Config.object().required(),
	onSelect: Config.func().value(noop)
};

class SearchableSelect extends Component {
	@autobind
	handleSelect(item) {
		this.props.onSelect(item);
	}

	render() {
		const {
			buttonPlaceholder,
			footerButtonMessage,
			footerOnClick,
			inputPlaceholder,
			inputValue,
			items,
			onInput,
			readOnly,
			selectedItem,
			showSearch
		} = this.props;

		return (
			<Dropdown
				{...this.otherProps()}
				elementClasses="searchable-select-root"
				label={(selectedItem && selectedItem.name) || buttonPlaceholder}
				readOnly={readOnly}
			>
				{showSearch && (
					<Dropdown.Section>
						<SearchInput
							onInput={onInput}
							placeholder={inputPlaceholder}
							value={inputValue}
						/>
					</Dropdown.Section>
				)}

				<Dropdown.Section elementClasses="items-wrapper">
					{items.map(
						(item, i) =>
							do {
								if (item.subheader) {
									<Dropdown.Subheader key={item.name}>
										{item.name}
									</Dropdown.Subheader>;
								} else {
									<Item
										active={
											selectedItem &&
											item.value === selectedItem.value
										}
										item={item}
										key={i}
										onSelect={this.handleSelect}
									/>;
								}
							}
					)}
				</Dropdown.Section>

				{footerOnClick && (
					<div class="footer-action">
						<Button block display="primary" onClick={footerOnClick}>
							{footerButtonMessage}
						</Button>
					</div>
				)}
			</Dropdown>
		);
	}
}

SearchableSelect.PROPS = {
	buttonPlaceholder: Config.string().value(''),
	footerButtonMessage: Config.string(),
	footerOnClick: Config.func(),
	inputPlaceholder: Config.string().value(''),
	inputValue: Config.string().value(''),
	items: Config.array().required(),
	onInput: Config.func().value(noop),
	onSelect: Config.func().value(noop),
	readOnly: Config.bool(),
	selectedItem: Config.shapeOf({
		name: Config.string(),
		value: Config.any()
	}),
	showSearch: Config.bool().value(true)
};

export default SearchableSelect;