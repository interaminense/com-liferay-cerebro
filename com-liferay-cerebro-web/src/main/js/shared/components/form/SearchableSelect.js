import Component, {Config} from './Component';
import getCN from 'classnames';
import {uniqueId} from 'lodash';

import SearchableSelect from '../SearchableSelect';
import Select from '../Select';

class FormSearchableSelect extends Component {
	created() {
		this._id = uniqueId('FormSearchableSelect');

		const {initialValue} = this.props;

		if (initialValue) {
			this.state.value = initialValue;
		}
	}

	syncInitialValue(newVal) {
		if (newVal) {
			this.state.value = newVal;
		}
	}

	render() {
		const {show, valid, value} = this.state;

		const classes = getCN({
			'has-danger': !valid,
			'has-success': valid && show
		});

		return (
			<SearchableSelect
				{...this.otherProps()}
				elementClasses={classes}
				inputPlaceholder={Liferay.Language.get('search')}
				selectedItem={value}
			/>
		);
	}
}

FormSearchableSelect.PROPS = {
	initialValue: Config.object()
};

FormSearchableSelect.STATE = {
	show: Config.value(false),
	valid: Config.value(true),
	value: Config.object()
};

FormSearchableSelect.Item = Select.Item;

export default FormSearchableSelect;