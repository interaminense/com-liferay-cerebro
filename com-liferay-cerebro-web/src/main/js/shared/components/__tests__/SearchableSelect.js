import dom from 'metal-dom';
import {times} from 'lodash';

import SearchableSelect from '../SearchableSelect';
import renderWithOverlay from 'test/overlay';

describe('SearchableSelect', () => {
	const items = times(5, i => ({name: `item${i}`, value: i}));

	let component;

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render', () => {
		component = renderWithOverlay(SearchableSelect, {
			buttonPlaceholder: 'Bar',
			inputPlaceholder: 'Foo',
			items
		});

		expect(component).toMatchSnapshot();
	});

	it('should render when clicked', () => {
		component = renderWithOverlay(SearchableSelect, {
			buttonPlaceholder: 'Bar',
			inputPlaceholder: 'Foo',
			items
		});

		dom.triggerEvent(
			component.element.querySelector('.dropdown-toggle'),
			'click'
		);

		jest.runAllTimers();

		expect(component).toMatchSnapshot();
	});

	it('should render with a subheader', () => {
		component = renderWithOverlay(SearchableSelect, {
			buttonPlaceholder: 'Bar',
			inputPlaceholder: 'Foo',
			items: [{name: 'foo bar', subheader: true}, ...items]
		});

		dom.triggerEvent(
			component.element.querySelector('.dropdown-toggle'),
			'click'
		);

		jest.runAllTimers();

		expect(component).toMatchSnapshot();
	});
});