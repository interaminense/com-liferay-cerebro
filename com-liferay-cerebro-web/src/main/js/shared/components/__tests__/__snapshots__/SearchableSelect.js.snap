// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchableSelect should render 1`] = `
<span>
  <div class="overlay-renderer-root"></div>
  <div class="dropdown dropdown-root searchable-select-root" id="overlay1">
    <button class="button-root btn btn-secondary dropdown-toggle" type="button">
      <span class="text-truncate">Bar </span>
      <span>
        <svg class="icon-root lexicon-icon lexicon-icon-caret-bottom">
          <use xlink:href="/images/lexicon/icons.svg#caret-bottom"></use>
        </svg>
      </span>
    </button>
  </div>
</span>
`;

exports[`SearchableSelect should render when clicked 1`] = `
<span>
  <div class="overlay-renderer-root">
    <div data-overlayid="2" id="overlay2Container" class="card overlay-container-root dropdown dropdown-root searchable-select-root" style="top: 0px; left: 0px;">
      <div class="dropdown-menu show">
        <div class="dropdown-section">
          <div class="">
            <div class="input-group input-group-inset">
              <div class="input-group-input">
                <input value="" placeholder="Foo" class="input-root form-control" type="text">
              </div>
              <span class="input-group-inset-item">
                <button class="button-root btn btn-unstyled" type="button">
                  <svg class="icon-root lexicon-icon lexicon-icon-search">
                    <use xlink:href="/images/lexicon/icons.svg#search"></use>
                  </svg>
                </button>
              </span>
            </div>
          </div>
        </div>
        <div class="dropdown-section items-wrapper">
          <button class="button-root btn btn-secondary dropdown-item text-truncate" type="button">item0</button>
          <button class="button-root btn btn-secondary dropdown-item text-truncate" type="button">item1</button>
          <button class="button-root btn btn-secondary dropdown-item text-truncate" type="button">item2</button>
          <button class="button-root btn btn-secondary dropdown-item text-truncate" type="button">item3</button>
          <button class="button-root btn btn-secondary dropdown-item text-truncate" type="button">item4</button>
        </div>
      </div>
    </div>
  </div>
  <div class="dropdown dropdown-root searchable-select-root" id="overlay2" active="">
    <button class="button-root btn btn-secondary dropdown-toggle" type="button">
      <span class="text-truncate">Bar </span>
      <span>
        <svg class="icon-root lexicon-icon lexicon-icon-caret-bottom">
          <use xlink:href="/images/lexicon/icons.svg#caret-bottom"></use>
        </svg>
      </span>
    </button>
  </div>
</span>
`;

exports[`SearchableSelect should render with a subheader 1`] = `
<span>
  <div class="overlay-renderer-root">
    <div data-overlayid="3" id="overlay3Container" class="card overlay-container-root dropdown dropdown-root searchable-select-root" style="top: 0px; left: 0px;">
      <div class="dropdown-menu show">
        <div class="dropdown-section">
          <div class="">
            <div class="input-group input-group-inset">
              <div class="input-group-input">
                <input value="" placeholder="Foo" class="input-root form-control" type="text">
              </div>
              <span class="input-group-inset-item">
                <button class="button-root btn btn-unstyled" type="button">
                  <svg class="icon-root lexicon-icon lexicon-icon-search">
                    <use xlink:href="/images/lexicon/icons.svg#search"></use>
                  </svg>
                </button>
              </span>
            </div>
          </div>
        </div>
        <div class="dropdown-section items-wrapper">
          <div class="dropdown-subheader">foo bar</div>
          <button class="button-root btn btn-secondary dropdown-item text-truncate" type="button">item0</button>
          <button class="button-root btn btn-secondary dropdown-item text-truncate" type="button">item1</button>
          <button class="button-root btn btn-secondary dropdown-item text-truncate" type="button">item2</button>
          <button class="button-root btn btn-secondary dropdown-item text-truncate" type="button">item3</button>
          <button class="button-root btn btn-secondary dropdown-item text-truncate" type="button">item4</button>
        </div>
      </div>
    </div>
  </div>
  <div class="dropdown dropdown-root searchable-select-root" id="overlay3" active="">
    <button class="button-root btn btn-secondary dropdown-toggle" type="button">
      <span class="text-truncate">Bar </span>
      <span>
        <svg class="icon-root lexicon-icon lexicon-icon-caret-bottom">
          <use xlink:href="/images/lexicon/icons.svg#caret-bottom"></use>
        </svg>
      </span>
    </button>
  </div>
</span>
`;
